core/vdbe/mod.rs

```diff
@@ -2667,6 +2679,183 @@ fn exec_if(reg: &OwnedValue, null_reg: &OwnedValue, not: bool) -> bool {
+            OwnedValue::Integer(i) => OwnedValue::Float(*i as f64),
+            OwnedValue::Float(f) => OwnedValue::Float(*f),
+            _ => OwnedValue::Float(0.0),
+        },
+        "INTEGER" => match value {
+            OwnedValue::Blob(b) => {
+                // Convert BLOB to TEXT first
+                let text = String::from_utf8_lossy(b);
+                cast_text_to_integer(&text)
+            }
+            OwnedValue::Text(t) => cast_text_to_integer(t),
+            OwnedValue::Integer(i) => OwnedValue::Integer(*i),
+            // A cast of a REAL value into an INTEGER results in the integer between the REAL value and zero
+            // that is closest to the REAL value. If a REAL is greater than the greatest possible signed integer (+9223372036854775807)
+            // then the result is the greatest possible signed integer and if the REAL is less than the least possible signed integer (-9223372036854775808)
+            // then the result is the least possible signed integer.
+            OwnedValue::Float(f) => {
+                let i = f.floor() as i128;
+                if i > i64::MAX as i128 {
+                    OwnedValue::Integer(i64::MAX)
+                } else if i < i64::MIN as i128 {
+                    OwnedValue::Integer(i64::MIN)
+                } else {
+                    OwnedValue::Integer(i as i64)
+                }
+            }
+            _ => OwnedValue::Integer(0),
+        },
+        "NUMERIC" => match value {
+            OwnedValue::Blob(b) => {
+                let text = String::from_utf8_lossy(b);
+                cast_text_to_numeric(&text)
+            }
+            OwnedValue::Text(t) => cast_text_to_numeric(t),
+            OwnedValue::Integer(i) => OwnedValue::Integer(*i),
+            OwnedValue::Float(f) => OwnedValue::Float(*f),
+            _ => value.clone(), // TODO probably wrong
+        },
+        _ => value.clone(),
+    }
+}
+
+/// For tables not declared as STRICT, the affinity of a column is determined by the declared type of the column, according to the following rules in the order shown:
+/// If the declared type contains the string "INT" then it is assigned INTEGER affinity.
+/// If the declared type of the column contains any of the strings "CHAR", "CLOB", or "TEXT" then that column has TEXT affinity. Notice that the type VARCHAR contains the string "CHAR" and is thus assigned TEXT affinity.
+/// If the declared type for a column contains the string "BLOB" or if no type is specified then the column has affinity BLOB.
+/// If the declared type for a column contains any of the strings "REAL", "FLOA", or "DOUB" then the column has REAL affinity.
+/// Otherwise, the affinity is NUMERIC.
+/// Note that the order of the rules for determining column affinity is important. A column whose declared type is "CHARINT" will match both rules 1 and 2 but the first rule takes precedence and so the column affinity will be INTEGER.
+fn affinity(datatype: &str) -> &str {
```

Comment:
Would it be better to return an Enum so that we don't compare strings and we compare numbers instead?

Reply:
yes it would

context switching between this and javascript...

Reply:
fixed